wa/instruments/perf.py

```diff
@@ -245,11 +238,160 @@ def stop(self, context):
+          among the events through time-slicing. This classifier holds the
+          fraction (between ``0.0`` and ``100.0``) of the run that a hardware
+          counter was allocated to the the event. Available only for hardware
+          events and only when time-slicing was required.
+        - ``'comment_value'``, ``'comment_units'``: Some counters may come with
+          an extra "comment" (following a ``#``) added by ``perf``. The
+          ``'comment_value'`` holds the numeric (``int`` or ``float``) value of
+          the comment while ``'comment_units'`` holds the rest of the comment
+          (typically the units). Only available for the events for which
+          ``perf`` added a comment.
+        """
+        match = cls._stat_regex.search(stdout)
+        if match is None:
+            return
+        base_classifiers = {
+            'label': label,
+            'target': match['target'],
+            'duration': float(match['duration'].replace(',', '')),
+            'duration_units': match['duration_units'],
+        }
+        for m in cls._stat_counter_regex.finditer(match['counters']):
+            classifiers = base_classifiers.copy()
+            name, count = cls._extract_stat_count(m, classifiers)
+            yield {
+                'name': name,
+                'units': None,
+                'value': count,
+                'classifiers': classifiers,
+            }
+
+    _stat_regex = re.compile(
+        r'Performance counter stats for (?P<target>.*?)\s*:\s*$'
+        r'^(?P<counters>.*)$'
+        r'^\s*(?P<duration>[0-9.,]+)\s*(?P<duration_units>\S+)\s*time elapsed',
+        flags=(re.S | re.M))
+
+    _stat_counter_regex = re.compile(
+        r'^\s*{aggr}?\s*{count}\s*{name}\s*{comment}?(?:{enabled}|$)'.format(
+            aggr=r'(?:{hw_thread}|(?:{cluster}{core}?\s*{thread_cnt}))'.format(
+                hw_thread=r'(?:CPU-?(?P<hw_thread>\d+))',
+                cluster=r'S(?P<cluster>\d+)',
+                core=r'(?:-C(?P<core>\d+))',
+                thread_cnt=r'(?P<hw_thread_count>\d+)'),
+            count=r'(?P<count>[0-9.,]+|\<not supported\>|\<not counted\>)',
+            name=r'(?P<name>.*?)',
+            comment=r'(?:#\s*{value}\s*{units}\s*)'.format(
+                value=r'(?P<comment_value>[0-9,.]+)',
+                units=r'(?P<comment_units>.*?)'),
+            enabled=r'(?:[\[\(](?P<enabled>[0-9.]+)%[\)\]])'),
+        flags=re.M)
+
+    @staticmethod
+    def _extract_stat_count(match, classifiers):
+        """Extracts the counter classifiers and count from a counter_match.
+
+        Parameters:
+            match        A :class:`re.Match` from :attr:`_stat_counter_regex`
+            classifiers  A dictionary to be completed for the matched counter
+
+        Returns:
+            A (name, value) tuple for the matched counter (value is 0 if an
+            error occurred).
+        """
+        name = f'{classifiers["label"]}_{match["name"]}'.replace(' ', '_')
```

Comment:
As per earlier discussion, we'd like to continue supporting Python 3.5 for the foreseeable future in order to avoid forcing potentially  painful migrations on existing WA users. Because of that, 3.6 features, such as format strings, should not be used in WA code.

Reply:
I wasn't sure what the conclusion of ARM-Software/devlib#389 was; I have changed the f-strings to `format` calls.

Reply:
Yeah, that discussion kinda petered out without a concrete conclusion. However, I assume that earlier issue regarding support for Ubuntu 16.04 stands, and since there is nothing _essential_ in 3.6, we're going to stick with 3.5 for the next release, at least.